### DISCUSSION TITLE
Exploring the clustering experiment of the AC BO Hackathon

### DISCUSSION THREAD

**Discussion by sgbaird** [emdgroup/baybe#283] [Transfer Learning]
There was surprising behavior from the corresponding notebook: https://github.com/AC-BO-Hackathon/project-26-multiple-context-bo/blob/main/cluster_experiment.ipynb

> The key result of this study is that choosing clusters instead of providing the full available data definitely has an influence on the performance and might be worth a more in-depth investigation. Interestingly, it seems like the baseline actually performs worst in our experiments, although it has the maximum number of points available. This could have several reasons, like the chosen function not being a suitable one for this exercise, or the availability of relatively many points distracting the optimizer. Also, it might be the case that it might take more iterations for the baseline to "catch up" and overtaking the experiments using clusters.

In particular, notice how the "baseline" (which only contains an initial 30 datapoints, no additional clusters)

![image](https://github.com/emdgroup/baybe/assets/45469701/5b1bfc60-de73-4cbf-adf5-99ec97b440a6)

Compared with the [transfer learning example from the BayBE docs](https://emdgroup.github.io/baybe/userguide/transfer_learning.html) [[full example](https://emdgroup.github.io/baybe/examples/Transfer_Learning/basic_transfer_learning.html)], the only difference is supposedly the use of clusters vs. the use of individual datapoints (and perhaps the amount of the data since clusters would have multiple data points (?)).

![image](https://github.com/emdgroup/baybe/assets/45469701/261d0779-10bc-4579-adc1-f84ab221807d)

**Comment by mehradans92** [comment]
@sgbaird Rerunning the notebook, I can't produce the same result. Compare lines for 2, 5 and 10.

![cluster_experiments_30_fallback](https://github.com/emdgroup/baybe/assets/51170839/c39aba43-3ef9-4c68-986a-45d2c1520837)

This is for 30 initial points.

**Reply by AVHopp** [reply]
I am getting a rather similar, but not identical plot with 0.9.0:
![output](https://github.com/emdgroup/baybe/assets/84016733/f47cc1ef-6c47-4f9c-80f1-ccdf81a02064)
With 0.9.1 I get the following: 
![image](https://github.com/emdgroup/baybe/assets/84016733/6ebc1849-9e45-4402-8c68-c14754067f74)
Not sure if there is undeterministic behavior involved?

**Comment by mehradans92** [comment]
Hi @AdrianSosic. Could you please let us know which version of `baybe` was used to generate this? Currently, I have `0.9.1` installed with `pip`.
I am seeing a much different behavior compared to what's in https://github.com/AC-BO-Hackathon/project-26-multiple-context-bo/blob/main/cluster_experiment.ipynb tho. See below:

Using 50 initial points:
![cluster_experiments_50_fallback](https://github.com/emdgroup/baybe/assets/51170839/7d49bdef-75d1-42ad-8a02-1b45b146ba0c)

Using 100 initial points:
![cluster_experiments_100_fallback](https://github.com/emdgroup/baybe/assets/51170839/d140b2ab-0c2f-4081-972b-e29294f1d828)

**Comment by AVHopp** [comment]
I did these experiments during the hackathon, so I'll have a look into this. One thing for sure is that it was not done with version 0.9.1 since that version was not out at that time. Looking at the versions, it was probably 0.8.1 or 0.8.2, I will investigate.

**Comment by AVHopp** [comment]
Some updates from my side:

- I had trouble running the notebook with any version `baybe<0.9.0`. It would be great if you could elaborate on whether you can create plots with lower BayBE versions. From the errors I see, the issue might well be just related to my local environments or packages that are unavailable/different for a fixed version of BayBE, probably because the tests were done with an "intermediate" version.
- Note that you cannot directly compare the performance in this notebook with the one in the examples: The notebook uses a *continuous* search space and clusters within that, while the examples that we have on the website show a *discretized* search space.
- However, there is in fact weird behavior here, and we (mainly @AdrianSosic and me) are currently investigating this. I'm currently running some tests with a fresh 0.9.0 installation of BayBE and will share my results here. Reasons for different behavior when using `baybe>=0.9.0` are probably some things that were changed in the backend in between.

Hope this makes some things a bit more clear. If not just ask, always happy to help 😄

**Reply by sgbaird** [reply]
> Note that you cannot directly compare the performance in this notebook with the one in the examples: The notebook uses a continuous search space and clusters within that, while the examples that we have on the website show a discretized search space.
@AVHopp ah, thanks for catching! I had assumed the cluster experiment was discrete like the website, but hadn't dug in enough to the notebook code to see the distinction. I've mostly been focused on the website example.

For the motivation behind continuous vs. discrete benchmarking, it would be great to get some more context at https://github.com/emdgroup/baybe/discussions/257#discussion-6752132.

**Reply by AVHopp** [reply]
I tried to give some explanation regarding how we use examples there. Do we want to take the discussion there and can we close this discussion then?

**Reply by sgbaird** [reply]
Thanks for clarifying! Yes, for this thread we can move to https://github.com/emdgroup/baybe/discussions/257#discussioncomment-9902287, but I think for the top-level discussion around the strange behavior with clustering experiments, it might make sense to keep this one open. Up to you though.

**Reply by AVHopp** [reply]
My question would be if we want to open a new, related discussion regarding "Behavior of BayBE for different versions" as this seems to be the point that is what causes the actual confusion here. If this is fine for everybody involved (in particular @mehradans92 ), I'd then just move the insights that we had from this discussion to a new one

**Reply by mehradans92** [reply]
Yeah. Sure :)

**Comment by AVHopp** [comment]
Here are the plots I got for `baybe 0.9.0":
30: 
![image](https://github.com/emdgroup/baybe/assets/84016733/7fa52369-c599-4f98-85ed-531bb3306bc3)
50:
![image](https://github.com/emdgroup/baybe/assets/84016733/c3d1f511-0192-40aa-a6cc-4cfcf0afc1d2)
100:
![image](https://github.com/emdgroup/baybe/assets/84016733/7df9e364-ff54-4b42-bc49-4caf875d44e1)
250:
![image](https://github.com/emdgroup/baybe/assets/84016733/6abd6af1-189d-4187-b0ef-ffa430d0e3f2)

**Comment by AVHopp** [comment]
A quick update on this:
- The "weird behavior" mentioned in one of the comments above is fixed by now - the issue was related to using `TaskParameter` in a purely continuous space, which is working by now.
- Also, we are currently implementing more benchmark functionalities: There is an open PR for bringing both Transfer Learning and non-TL benchmarks to BayBE. These can be run locally, but are also used internally for evaluation of BayBE. The (currently open) PR related to this is #491 , and we expect this to be merged soon. So for benchmarking nowadays, we highly recommend to use the new `benchmarks` module and run this locally.
